feats/storage.py

```diff
@@ -11,9 +13,28 @@
 If this Storage has not yet seen a key, it should return an empty sequence.
 """
 
-Memory: Storage = lambda: defaultdict(list)
+Memory: Storage = lambda: defaultdict(MemoryList)
 """
 The Memory storage keeps all data in memory. When the application exits, all data
 will be lost.
 Mainly useful for testing environments.
 """
+class MemoryList(list):
+    """
+    An implementation of list with deepcopy guard rails to prevent
+    mutating state in the memory store
+    """
+    def last(self):
+        return self[-1]
+
+    def __getitem__(self, index):
+        item = super().__getitem__(index)
+        return deepcopy(item)
+
+    def __setitem__(self, i, value):
```

Comment:
Does it make more sense to actually have this method just raise an exception since setting a key is not a valid use case for our streams?

Reply:
Yeah that sounds good to me